File: packages/sim-core/src/production.ts
import type { AgentType, SimConfig } from "./types.js";

const EPS = 1e-9;

/**
 * CES aggregator: A [ α K^ρ + (1-α) L^ρ ]^(1/ρ).
 * ρ → 0 uses Cobb–Douglas limit K^α L^(1-α).
 */
export function cesAggregate(
  knowledge: number,
  labor: number,
  alpha: number,
  rho: number,
  scale: number,
): number {
  const K = Math.max(EPS, knowledge);
  const L = Math.max(EPS, labor);
  const a = Math.min(1, Math.max(0, alpha));
  const r = rho;
  if (Math.abs(r) < 1e-10) {
    return scale * Math.pow(K, a) * Math.pow(L, 1 - a);
  }
  const inner = a * Math.pow(K, r) + (1 - a) * Math.pow(L, r);
  return scale * Math.pow(Math.max(EPS, inner), 1 / r);
}

/** Type tilt: how labor splits between goods vs services capacity (per actor kind). */
export function serviceLaborShare(agentType: AgentType): number {
  switch (agentType) {
    case "academic":
      return 0.72;
    case "solo":
      return 0.55;
    case "smb":
      return 0.42;
    case "bigco":
      return 0.28;
    default:
      return 0.45;
  }
}

/**
 * Branch qualities and blended offering quality (used for CES revenue and regulatory externality channels).
 */
export function offeringQualityBranches(
  agent: {
    type: AgentType;
    knowledge: number;
    labor: number;
  },
  cfg: SimConfig,
): { qGood: number; qServ: number; q: number } {
  if (!cfg.cesQualityEnabled) {
    return { qGood: 1, qServ: 1, q: 1 };
  }
  const s = serviceLaborShare(agent.type);
  const L = Math.max(0, agent.labor);
  const Lg = L * (1 - s);
  const Ls = L * s;
  const qGood = cesAggregate(
    agent.knowledge,
    Lg,
    cfg.cesAlphaKnowledge,
    cfg.cesRho,
    cfg.cesScale,
  );
  const qServ = cesAggregate(
    agent.knowledge,
    Ls,
    cfg.cesAlphaKnowledge,
    cfg.cesRho,
    cfg.cesScale,
  );
  const q = cfg.cesMixGoods * qGood + (1 - cfg.cesMixGoods) * qServ;
  return { qGood, qServ, q };
}

/**
 * Effective offering quality: weighted CES for goods and services (same K; labor split).
 * Higher knowledge and appropriately allocated labor raise quality.
 */
export function offeringQuality(agent: {
  type: AgentType;
  knowledge: number;
  labor: number;
}, cfg: SimConfig): number {
  return offeringQualityBranches(agent, cfg).q;
}

File: packages/ip-sim-engine/src/sim/production.rs
//! Matches `packages/sim-core/src/production.ts`.

use crate::sim::types::{AgentKind, SimConfig};

const EPS: f64 = 1e-9;

pub fn ces_aggregate(
    knowledge: f64,
    labor: f64,
    alpha: f64,
    rho: f64,
    scale: f64,
) -> f64 {
    let k = knowledge.max(EPS);
    let l = labor.max(EPS);
    let a = alpha.clamp(0.0, 1.0);
    let r = rho;
    if r.abs() < 1e-10 {
        return scale * k.powf(a) * l.powf(1.0 - a);
    }
    let inner = a * k.powf(r) + (1.0 - a) * l.powf(r);
    scale * inner.max(EPS).powf(1.0 / r)
}

pub fn service_labor_share(kind: AgentKind) -> f64 {
    match kind {
        AgentKind::Academic => 0.72,
        AgentKind::Solo => 0.55,
        AgentKind::Smb => 0.42,
        AgentKind::Bigco => 0.28,
    }
}

pub struct BranchQuality {
    pub q_good: f64,
    pub q_serv: f64,
    pub q: f64,
}

pub fn offering_quality_branches(
    kind: AgentKind,
    knowledge: f64,
    labor: f64,
    cfg: &SimConfig,
) -> BranchQuality {
    if !cfg.ces_quality_enabled {
        return BranchQuality {
            q_good: 1.0,
            q_serv: 1.0,
            q: 1.0,
        };
    }
    let s = service_labor_share(kind);
    let l = labor.max(0.0);
    let lg = l * (1.0 - s);
    let ls = l * s;
    let q_good = ces_aggregate(
        knowledge,
        lg,
        cfg.ces_alpha_knowledge,
        cfg.ces_rho,
        cfg.ces_scale,
    );
    let q_serv = ces_aggregate(
        knowledge,
        ls,
        cfg.ces_alpha_knowledge,
        cfg.ces_rho,
        cfg.ces_scale,
    );
    let q = cfg.ces_mix_goods * q_good + (1.0 - cfg.ces_mix_goods) * q_serv;
    BranchQuality {
        q_good,
        q_serv,
        q,
    }
}
